docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+
+Thus, the source of truth for instance scoped variables moves from the `config` package to another part of the server (likely `domain`). See [issue #30366](https://github.com/pingcap/tidb/issues/30366).
+
+## Impacts & Risks
+
+The biggest risk is that we can not agree on a reduced scope of implementation and the project extends to cover increased scope. Configuration management is a classic example of a [bikeshed problem](https://en.wikipedia.org/wiki/Law_of_triviality), and we will likely need to make some tradeoffs to get anywhere.
+
+Many of the alternatives are difficult to implement because they break compatibility (we need to live with `GLOBAL` means cluster `GLOBAL`, and there is no `SET CLUSTER`) or they break rolling upgrade scenarios because we currently have no specific rules of the minimum version of TiDB which can be upgraded from.
+
+## Investigation & Alternatives
+
+- There is no equivalent functionality to reference in MySQL since it does not have the native concept of a cluster.
+- An [earlier proposal](https://docs.google.com/document/d/1RuajYAFVsjJCwCBpIYF--9jtgxDZZcz3cbP-iVuLxJI/edit) for `INSTANCE` scope, with rules for precedence (same author)
+- CockroachDB has cluster level settings and node-level settings. Most settings are cluster level, and [node-level](https://www.cockroachlabs.com/docs/v21.2/cockroach-start) needs to be parsed as arguments when starting the server.
+
+## Unresolved Questions
+
+There are **very few** known scenarios where a system variable is *currently both* instance and global scoped, but these individually would need to be handled:
+
+```
+package main
+
+import (
+	"fmt"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/pingcap/tidb/sessionctx/variable"
+)
+
+func main() {
+	for _, v := range variable.GetSysVars() {
+		if v.HasGlobalScope() && v.HasSessionScope() && (v.GetSession != nil || v.SetSession != nil) && (v.GetGlobal != nil || v.SetGlobal != nil) {
+			fmt.Printf("%s\n", v.Name)
+		}
+	}
+}
+```
+
+**Output:**
+
+```
+tidb_store_limit
+tidb_stmt_summary_internal_query
+tidb_enable_stmt_summary
+tidb_stmt_summary_max_sql_length
+tidb_stmt_summary_max_stmt_count
+tidb_capture_plan_baselines
+tidb_stmt_summary_refresh_interval
+tidb_stmt_summary_history_size
+```
+
+The following suggestions are provided (to be discussed):
+
+- `tidb_store_limit`: Suggestion is to convert to global only (Currently does not work correctly, see [issue #30515](https://github.com/pingcap/tidb/issues/30515))
+- `tidb_stmt_summary_XXX`: Suggestion is to convert to global only.
+- `tidb_enable_stmt_summary`: Convert to instance scope only.
```

Comment:
I'm OK with `tidb_stmt_summary_XXX` but concerned about `tidb_enable_stmt_summary`. 
- In v4.0 and later versions, `tidb_enable_stmt_summary` is `true` by default and users rarely change it. The only case I saw is to tune the performance. In this case, it's more convenient to set it globally.
- If `tidb_enable_stmt_summary` was `false` globally, the expected value is still `false` after upgrading. However, it will be `true` if we only support instance scope because we won't read the global system variable anymore.
- We do need an instance scope when it's disabled by default and we want to check the statements summary in one single instance to minimize the impact on performance. But I think it's also acceptable to enable it globally because it's recommended in current versions.
- The modification affects TiDB Dashboard because TiDB Dashboard sends requests when the user enables or disables the feature from TiDB Dashboard. /cc @breeswish @crazycs520

Reply:
I saw it as we have two options:

1. We make `tidb_enable_stmt_summary` instance and `ON` by default, and users can edit their configuration file (on a per instance basis) to disable it? Still compatible with TiDB Dashboard. Downside: because instance config doesn't persist, we can't create a perfect upgrade scenario because we can't use their current value as the default. It has to be compiled default.
2. We make `tidb_enable_stmt_summary` GLOBAL only. The downside is users can't use it to observe part of their cluster (assuming their is overhead).

I had assumed (1) was better, but it is by no means a requirement. Eason's preference is to use global variables instead of instance, so we can certainly change it to (2).

Reply:
My option is (2). When we set `set global tidb_enable_stmt_summary`, it affects the whole cluster.

Reply:
Yes, when user controls it in TiDB Dashboard, it should take effect globally.

Reply:
OK that's unanimous. I will update it.